File: Frontend/src/types/contract.ts
export type ContractStatus = 
  | 'pending' 
  | 'signed_farmer' 
  | 'active' 
  | 'completed' 
  | 'cancelled' 
  | 'rejected';

export interface Signature {
  signer_id: string;
  signer_name: string;
  signer_role: string;
  signature: string;
  signed_at: string;
  public_key: string;
}

export interface ContractItem {
  product_id: string;
  product_name: string;
  quantity: number;
  unit: string;
  price_per_unit: number;
  total_price: number;
}

export interface Contract {
  _id: string;
  buyer_id: string;
  buyer_name: string;
  farmer_id: string;
  farmer_name: string;
  items: ContractItem[];
  total_amount: number;
  delivery_date?: string;
  delivery_address?: string;
  terms?: string;
  notes?: string;
  status: ContractStatus;
  contract_hash: string;
  blockchain_tx_id?: string;
  farmer_signature?: Signature;
  buyer_signature?: Signature;
  created_at: string;
  updated_at: string;
  signed_at?: string;
  completed_at?: string;
}

export interface CreateContractRequest {
  buyer_id: string;
  buyer_name: string;
  farmer_id: string;
  farmer_name: string;
  items: ContractItem[];
  total_amount: number;
  delivery_date?: string;
  delivery_address?: string;
  terms?: string;
  notes?: string;
}

export interface SignContractRequest {
  signature: string;
  public_key: string;
}

export interface KeyPair {
  private_key: string;
  public_key: string;
  note: string;
}

File: Backend/app/schemas/contract.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class ContractItemSchema(BaseModel):
    product_id: str
    product_name: str
    quantity: float
    unit: str
    price_per_unit: float
    total_price: float


class SignatureSchema(BaseModel):
    signer_id: str
    signer_name: str
    signer_role: str
    signature: str
    signed_at: datetime
    public_key: str


class ContractCreate(BaseModel):
    buyer_id: str
    buyer_name: str
    farmer_id: str
    farmer_name: str
    items: List[ContractItemSchema]
    total_amount: float
    delivery_date: Optional[str] = None
    delivery_address: Optional[str] = None
    terms: Optional[str] = None
    notes: Optional[str] = None


class ContractResponse(BaseModel):
    _id: str
    buyer_id: str
    buyer_name: str
    farmer_id: str
    farmer_name: str
    items: List[ContractItemSchema]
    total_amount: float
    delivery_date: Optional[str] = None
    delivery_address: Optional[str] = None
    terms: Optional[str] = None
    notes: Optional[str] = None
    status: str
    contract_hash: str
    blockchain_tx_id: Optional[str] = None
    farmer_signature: Optional[SignatureSchema] = None
    buyer_signature: Optional[SignatureSchema] = None
    created_at: datetime
    updated_at: datetime
    signed_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    class Config:
        populate_by_name = True


class SignContractRequest(BaseModel):
    signature: str
    public_key: str


class KeyPairResponse(BaseModel):
    private_key: str
    public_key: str
    note: str
